accepted/2024/trusted-publishers-oidc-for-nuget-push.md

```diff
@@ -0,0 +1,523 @@
+
+In written form, this happens:
+
+1. The `nuget/login` workflow step does this:
+   - Fetch a GitHub Actions OIDC token, enabled via the `id-token: write` workflow permission
+   - Send the OIDC token to a NuGet.org endpoint, along with the NuGet.org username
+1. NuGet.org does this:
+   - Receive the OIDC token, verify the token using the public GitHub Actions keyset and other JWT rules
+   - Verify the provided username has a trust policy matching the OIDC token
+   - Generate a short-lived API key and return it to the caller
+1. The `nuget/login` workflow step does this:
+   - Receive the short-lived API key and set it as a step output
+2. The `dotnet nuget push` workflow step does this:
+   - Use the provided API key to perform a package push, using the [`PackagePublish`
+     resource](https://learn.microsoft.com/en-us/nuget/api/package-publish-resource)
+3. NuGet.org does this:
+   - Verify the provided package and package operation match the scopes on the API key
+   - Publish the package!
+
+## Technical explanation
+
+See the [technical document](trusted-publishers-oidc-for-nuget-push.technical.md) in this case directory for more
+information.
+
+## Drawbacks
+
+This workflow only helps users who want to build and push their packages on a recognized, Trusted Publisher. Not all
+CI/CD systems will be supported so some users will still need to use the older authentication flows such as long-lived,
+manually rotated API keys.
+
+The trust policy is configured once and works forever. An API key forces the package author to re-assess CI/CD
+infrastructure on a regular basis. We can consider a ["dead man's
+switch"](https://en.wikipedia.org/wiki/Dead_man%27s_switch) approach to force the package owner to attest that the trust
+policy is still desired if this is a big concern.
+
+If a series of pushes in a workflow outlive a short-lived API keys, later pushes may fail with "API key expired" errors.
+This could be mitigated by creating a NuGet credential provider that fetches the API key during the push operation, if
+it detects the current API key is expired.
+
+Because a trust policy is defined at the user level, there can be a mismatch of user or organization profile the trust
+policy is acting on behalf of. This is disambiguated by the "package owner" scope on the trust policy, much like the
+existing long-lived API key flow, but this can still be confusing. If we find this flow doesn't work well for users, an
+additional parameter can be added to the `nuget/login` step which is `owner`. If provided, this will assert that the
+desired package owner is `X` (such as an organization name). If no trust policy is found on the user for that package
+owner, a helpful error could be provided. We can wait on implementing this until we hear more from users.
+
+## Rationale and alternatives
+
+### Naming
+
+The name "trusted publishers" may seem overly vague. There has been conversations about this [on Hacker
```

Comment:
💯 the name is so confusing but for consistency's sake may be best to use it or reference it regularly.